src/main/java/cubicchunks/world/dependency/DependencyManager.java

```diff
@@ -0,0 +1,187 @@
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+
+package cubicchunks.world.dependency;
+
+import cubicchunks.CubicChunks;
+import cubicchunks.server.ServerCubeCache;
+import cubicchunks.server.ServerCubeCache.LoadType;
+import cubicchunks.util.CubeCoords;
+import cubicchunks.world.cube.Cube;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+ * TODO: Commenting
+ */
+public class DependencyManager {
+
+	// TODO: Make private.
+	public ServerCubeCache cubeProvider;
+
+	private Map<CubeCoords, RequiredCube> requiredMap;
+
+
+	public DependencyManager(ServerCubeCache cubeProvider) {
+		this.cubeProvider = cubeProvider;
+		this.requiredMap = new HashMap<>();
+	}
+
+
+	/**
+	 * Registers a given Dependent. All of the cubes it requires will be loaded up to the stage it needs them to be
+	 * at and it will be notified whenever one of its required cubes advances a stage.
+	 *
+	 * @param dependent The Dependent to be registered.
+	 */
+	public void register(Dependent dependent) {
+
+		// If the dependent does not have any requirements, there is nothing left to do here.
+		Collection<Requirement> requirements = dependent.getRequirements();
+
+		// Add all of the requirements.
+		for (Requirement requirement : requirements) {
+
+			// If the required cube is required by other dependents already, retrieve its RequiredCube instance,
+			// otherwise create a new one.
+			RequiredCube requiredCube = this.requiredMap.get(requirement.getCoords());
+			if (requiredCube == null) {
+
+				// If the required cube is loaded already, add it to the RequiredCube instance immediately.
+				Cube cube = this.cubeProvider.getCube(requirement.getCoords());
+				if (cube != null) {
+					requiredCube = new RequiredCube(cube);
+				}
+				// Otherwise create an empty instance.
+				else {
+					requiredCube = new RequiredCube();
+				}
+
+				this.requiredMap.put(requirement.getCoords(), requiredCube);
+			}
+			// Make sure the required cube's target stage is correct.
+			else {
+				requiredCube.setTargetStage(requirement.getTargetStage());
+			}
+
+			// Add the dependent to the required cubes list of dependents.
+			requiredCube.addDependent(dependent);
+
+			// If the required cube is loaded, notify the dependent.
+			if (requiredCube.isAvailable()) {
+				dependent.update(this, requiredCube.getCube());
+
+				// Has the cube reached its target stage?
+				if (requiredCube.getCube().getCurrentStage() == requiredCube.getCube().getTargetStage()) {
```

Comment:
As of right now it cannot be after its target stage unless something went wrong. The only component advancing stages is the GeneratorPipeline, which does not advance past the target stage. However, using your method is cleaner and I will change it.